$NetBSD: patch-ai,v 1.5 2025/10/12 00:57:39 mrg Exp $

- clean up junk decls that exist in headers, and include the right headers
- use fgets instead of gets
- fix printf to always have format
- when executing a child metamail, propagate the parents options as much as
  possible
- use termios instead of termio/sgtty
- give many things return types and (basic) prototypes for GCC 14.

--- metamail/metamail.c.orig	1994-02-16 17:57:19.000000000 -0800
+++ metamail/metamail.c	2025-10-11 16:45:09.001504426 -0700
@@ -26,6 +26,8 @@ WITHOUT ANY EXPRESS OR IMPLIED WARRANTIE
 #include <config.h>
 #include <patchlevel.h>
 
+#include "metamail.h"
+
 #ifdef BORLAND
 #define F_OK 0
 extern unsigned _stklen = 16384;
@@ -46,12 +48,8 @@ extern char *mktemp(char *);
 #include <signal.h>
 
 #ifndef AMIGA
-#ifdef SYSV
-#include <termio.h>
+#include <termios.h>
 #include <unistd.h>
-#else /* SYSV */
-#include <sgtty.h>
-#endif /* SYSV */
 #endif /* AMIGA */
 #endif /* MICROSOFT */
 #endif /* BORLAND */
@@ -83,7 +81,7 @@ extern char *MkRmScript();
 #define MAX_FILE_NAME_SIZE 256
 #define WRITE_BINARY	"w"
 #else /* AMIGA */
-extern char **environ, *gets();
+extern char **environ;
 #define CATCOMMAND  "cat"
 #define CATTEMPLATE "cat %s"
 #define METAMAIL    "metamail"
@@ -100,15 +98,49 @@ extern char **environ, *gets();
 #define CMDSIZE 1200 /* Maximum size of command to execute */
 
 #define LINE_BUF_SIZE       2000
-#ifndef MICROSOFT
-extern char *malloc();
-extern char *realloc();
-#endif
-extern char *getenv();
-extern char *index();
-extern char *rindex();
+#include <string.h>
+#include <stdlib.h>
+
 char fileToDelete[MAX_FILE_NAME_SIZE];
 
+void PauseForUser(void);
+void maybephead(char *);
+int ProcessArguments(int, char **);
+int ExitWithError(char *);
+void RestoreTtyState();
+int HandleMessage(char *, int);
+int Read822Prefix(int, int);
+int ProcessMailcapFiles(char *);
+int SaveSquirrelFile(char *);
+void PrepareMessage();
+void MkTmpFileName(char *);
+int TryBuiltIns(char *);
+void TranslateInputToOutput();
+void usage();
+void StripTrailingSpace(char *);
+void RunInNewWindow();
+void CreateNewWindowPrefix();
+void SetUpEnvironment();
+int ProcessMailcapFile();
+int TryMailcapEntry();
+int IsDirectory();
+int CtypeMatch();
+int GetMailcapEntry();
+int ExecuteMailcapEntry();
+int OKToRun();
+int ExecuteCommand();
+int PassesTest();
+void BuildCommand();
+int StartRawStdin();
+void strcatquoting();
+int NeedToBeQuiet();
+int NeedToAskBeforeExecuting();
+void SaveTtyState();
+int WriteTmpFile();
+void EliminateNastyChars();
+void strcpynoquotes();
+void phead();
+
 char *FindParam();
 extern FILE *popen();
 static char *nomem = "Out of memory!";
@@ -151,6 +183,32 @@ int MightAskBeforeExecuting = 1,
     JustWriteFiles = 0,
     ProcessingErrors = 0;
 
+static const char *
+makeoptions(void)
+{
+    static char buf[1024];
+    buf[0] = 0;
+    if (!MightAskBeforeExecuting)
+	strlcat(buf, " -d", sizeof(buf));
+    if (DefinitelyNotTty && MustNotBeTty)
+	strlcat(buf, " -x", sizeof(buf));
+    if (MaybePageOutput)
+	strlcat(buf, MustPageOutput ? " -P" : " -p", sizeof(buf));
+    if (EatLeadingNewlines)
+	strlcat(buf, " -e", sizeof(buf));
+    if (!PrintSomeHeaders && Quiet)
+	strlcat(buf, " -q", sizeof(buf));
+    if (DoInBackground)
+	strlcat(buf, " -B", sizeof(buf));
+    if (!Is822Format)
+	strlcat(buf, " -b", sizeof(buf));
+    if (PrintingMode)
+	strlcat(buf, " -h", sizeof(buf));
+    if (JustWriteFiles)
+	strlcat(buf, " -w", sizeof(buf));
+    return buf;
+}
+
 char *ContentType = NULL,
     *ContentEncoding = NULL,
     *MailerName = "unknown",
@@ -295,6 +353,7 @@ char *auxpath;
     }
 }
 
+void
 main(argc, argv)
 int argc;
 char **argv;
@@ -387,6 +446,7 @@ char *fname;
     }
 }
 
+int
 HandleMessage(SquirrelFile, nestingdepth)
 char *SquirrelFile;
 /* SquirrelFile, if non-NULL, is a place to save a recognized body instead of executing it. */
@@ -497,8 +557,9 @@ int nestingdepth;
         if (IsAlternative) {
             if (WroteSquirrelFile) {
                 int retcode; 
-                char Cmd[TMPFILE_NAME_SIZE + 15];
-                sprintf(Cmd, "%s %s", METAMAIL, NewSquirrelFile);
+                char Cmd[TMPFILE_NAME_SIZE + 15 + 1024];
+                snprintf(Cmd, sizeof(Cmd), "%s %s %s", METAMAIL, makeoptions(),
+		    NewSquirrelFile);
                 fflush(stdout); fflush(stderr);
                 retcode = system(Cmd);
 #ifdef MSDOS
@@ -579,7 +640,7 @@ int nestingdepth;
                                 int overwriteans = -1;
                                 do {
                                     printf("File %s exists.  Do you want to overwrite it (y/n) ?\n", Fname);
-                                    s = gets(AnsBuf);
+				    s = fgets(AnsBuf, sizeof(AnsBuf), stdin);
                                     if (!s) {
                                         overwriteans = 0;
                                     } else {
@@ -661,6 +722,7 @@ int nestingdepth;
     return(-1); /* Unrecognized, really */
 }
 
+int
 ProcessArguments(argc, argv)
 int argc;
 char **argv;
@@ -940,11 +1002,13 @@ char **argv;
     return(0);
 }
 
+void
 usage() {
     fprintf(stderr, "Usage:  metamail [-b] [-B] [-d] [-e] [-h] [-r] [-R] [-p]  [-P] [-x] [-y] [-z] [-c content-type] [-E content-transfer-encoding] [-f from-name] [-m mailername] [-s subject] [message-file-name]\n");
     ExitWithError(NULL);
 }
 
+void
 RunInNewWindow(argv, argc, SourceFileNamePtr)
 char **argv, **SourceFileNamePtr;
 int argc;
@@ -1026,6 +1090,7 @@ struct MailcapEntry BuiltInsAlternative[
     {"text/plain", CATTEMPLATE, NULL, 0, 1, 0, "plain text", LPRTEMPLATE},
     {NULL, NULL, NULL, 0, 0, 0}};
 
+int
 ProcessMailcapFiles(SquirrelFile) 
 char *SquirrelFile;
 {
@@ -1082,6 +1147,7 @@ char *SquirrelFile;
 #endif /* MICROSOFT */
 }
 
+int
 TryBuiltIns(SquirrelFile) 
 char *SquirrelFile;
 {
@@ -1094,6 +1160,7 @@ char *SquirrelFile;
     return(-1);
 }
 
+int
 ProcessMailcapFile(file, SquirrelFile)
 char *file, *SquirrelFile;
 {
@@ -1166,6 +1233,7 @@ eatmore:
     }
 }    
 
+int
 TryMailcapEntry(mc, SquirrelFile)
 struct MailcapEntry mc;
 char *SquirrelFile;
@@ -1189,6 +1257,7 @@ char *SquirrelFile;
     return(-1);
 }
 
+int
 SaveSquirrelFile(SquirrelFile)
 char *SquirrelFile;
 {
@@ -1202,9 +1271,9 @@ char *SquirrelFile;
     fprintf(outfp, "Content-type: %s", ContentType);
     for (j=0; j<CParamsUsed; ++j) {
         fprintf(outfp, " ; ");
-        fprintf(outfp, CParams[j]);
+        fprintf(outfp, "%s", CParams[j]);
         fprintf(outfp, " = ");
-        fprintf(outfp, CParamValues[j]);
+        fprintf(outfp, "%s", CParamValues[j]);
     }
     fprintf(outfp, "\n\n"); 
     TranslateInputToOutput(InputFP, outfp, EncodingCode, ContentType);
@@ -1214,6 +1283,7 @@ char *SquirrelFile;
     return(0);
 }
 
+int
 ExecuteMailcapEntry(mc, TmpFileName, ThisContentType)
 char *TmpFileName, *ThisContentType;
 struct MailcapEntry mc;
@@ -1361,6 +1431,7 @@ struct MailcapEntry mc;
     return(0);
 }
 
+int
 PassesTest(mc)
 struct MailcapEntry *mc;
 {
@@ -1414,6 +1485,7 @@ char *s, **t;
     return(NULL);
 }	
 
+int
 GetMailcapEntry(fp, mc)
 FILE *fp;
 struct MailcapEntry *mc;
@@ -1507,6 +1579,7 @@ struct MailcapEntry *mc;
     return(1);
 }
 
+int
 ExitWithError(txt)
 char *txt;
 {
@@ -1534,6 +1607,7 @@ char *s;
     return(newcopy);
 }
 
+int
 Read822Prefix(PrintHeads, nestingdepth)
 int PrintHeads, nestingdepth;
 {
@@ -1654,6 +1728,7 @@ yankagain:
     return 0;
 }
 
+void
 PrepareMessage() {
     int c;
 
@@ -1688,6 +1763,7 @@ PrepareMessage() {
     SetUpEnvironment();  
 }
 
+void
 SetUpEnvironment() { 
     int i, j, environsize;
     char **newenviron, *mailervar, *summaryvar, *ctypevar, *s;
@@ -1809,6 +1885,7 @@ char *def;
 }
 #endif
 
+int
 OKToRun(ctype, progname, label)
 char *ctype, *progname, *label;
 {
@@ -1823,7 +1900,7 @@ char *ctype, *progname, *label;
         } else {
             printf("This message contains '%s'-format data.\nDo you want to view it using the '%s' command (y/n) [y] ? ", ctype, ShortCommand(progname));
         }
-        s = gets(AnsBuf);
+	s = fgets(AnsBuf, sizeof(AnsBuf), stdin);
         if (!s) return(0); /* EOF */
 	while (s && *s && isspace((unsigned char) *s)) ++s;
 	if (*s == 'y' || *s == 'Y' || !*s || *s == '\n') return(1);
@@ -1834,6 +1911,7 @@ char *ctype, *progname, *label;
     }
 }
 
+void
 EliminateNastyChars(s)
 char *s;
 {
@@ -1848,6 +1926,7 @@ char *s;
     }
 }
 
+void
 StripTrailingSpace(s)
 char *s;
 {
@@ -1900,7 +1979,7 @@ char *hdr;
 }
 
 /* check the header given to see if it matches any in the KeyHeadList */
-maybephead(hdr)
+void maybephead(hdr)
 char *hdr;
 {
     char *s;
@@ -1934,6 +2013,7 @@ char *hdr;
 }
 
 /* This next routine prints out a mail header, and needs to deal with the new extended charset headers. */
+void
 phead(s)
 char *s;
 {
@@ -2022,7 +2102,8 @@ int ShowLeadingWhitespace;
     if (lc2strcmp(charset, PrevCharset)) {
         char *s2, *charsetinuse;
 
-        strcpy(PrevCharset, charset);
+        strncpy(PrevCharset, charset, sizeof(PrevCharset));
+	PrevCharset[sizeof(PrevCharset) - 1] = '\0';
         for (s2=PrevCharset; *s2; ++s2) {
             if (isupper((unsigned char) *s2)) *s2 = tolower((unsigned char) *s2);
         }
@@ -2032,7 +2113,7 @@ int ShowLeadingWhitespace;
         }
     }
     if (ecode == ENCODING_NONE) {
-        printf(txt+1);
+        printf("%s", txt+1);
     } else {
         /* What follows is REALLY bogus, but all my encoding stuff is pipe-oriented right now... */
         MkTmpFileName(TmpFile);
@@ -2067,6 +2148,7 @@ int ShowLeadingWhitespace;
     PrintHeader(txtend + 2, 0);
 }
 
+void
 BuildCommand(Buf, controlstring, TmpFileName, UsedTmpFileName, PartsWritten)
 char *Buf, *controlstring, *TmpFileName;
 int *UsedTmpFileName;
@@ -2279,6 +2361,7 @@ struct part **PartsWritten;
     *to = 0;
 }
 
+void
 strcpynoquotes(t,f)
 char *t, *f;
 {
@@ -2290,6 +2373,7 @@ char *t, *f;
     }
 }
 
+int
 WriteTmpFile(fname, ctype)
 char *fname;
 char *ctype;
@@ -2309,6 +2393,7 @@ char *ctype;
 }
 
 
+void
 TranslateInputToOutput(InputFP, OutputFP, Ecode, ctype)
 FILE *InputFP, *OutputFP;
 int Ecode;
@@ -2336,6 +2421,7 @@ char *ctype;
 #endif
 }
 
+void
 CreateNewWindowPrefix(Prefix)
 char *Prefix;
 {
@@ -2375,43 +2461,30 @@ char *Prefix;
 
 int HasSavedTtyState=0;
 #if !defined(AMIGA) && !defined(MSDOS)
-#ifdef SYSV
-static struct termio MyTtyStateIn, MyTtyStateOut;
-#else
-static struct sgttyb MyTtyStateIn, MyTtyStateOut;
-#endif
+static struct termios MyTtyStateIn, MyTtyStateOut;
 #endif
 
+void
 SaveTtyState() {
     /* Bogus -- would like a good portable way to reset the terminal state here */
 #if !defined(AMIGA) && !defined(MSDOS)
-#ifdef SYSV
-    ioctl(fileno(stdin), TCGETA, &MyTtyStateIn);
-    ioctl(fileno(stdout), TCGETA, &MyTtyStateOut);
-#else
-    gtty(fileno(stdin), &MyTtyStateIn);
-    gtty(fileno(stdout), &MyTtyStateOut);
-#endif
+    tcgetattr(fileno(stdin), &MyTtyStateIn);
+    tcgetattr(fileno(stdout), &MyTtyStateOut);
     HasSavedTtyState = 1;
 #endif
 }
 
+void
 RestoreTtyState() {
 #if !defined(AMIGA) && !defined(MSDOS)
-#ifdef SYSV
     if (HasSavedTtyState) {
-        ioctl(fileno(stdout), TCSETA, &MyTtyStateOut);
-        ioctl(fileno(stdin), TCSETA, &MyTtyStateIn);
+        tcsetattr(fileno(stdout), TCSANOW, &MyTtyStateOut);
+        tcsetattr(fileno(stdin), TCSANOW, &MyTtyStateIn);
     }
-#else
-    if (HasSavedTtyState) {
-        stty(fileno(stdout), &MyTtyStateOut);
-        stty(fileno(stdin), &MyTtyStateIn);
-    }
-#endif
 #endif
 }
 
+int
 NeedToAskBeforeExecuting(type)
 char *type;
 {
@@ -2423,6 +2496,7 @@ char *type;
     return(1);
 }
 
+int
 NeedToBeQuiet(cmd)
 char *cmd;
 {
@@ -2433,6 +2507,7 @@ char *cmd;
     return(0);
 }
 
+int
 CtypeMatch(ctype, pat)
 char *ctype, *pat;
 {
@@ -2459,6 +2534,7 @@ char *ctype, *pat;
     return(0);
 }
 
+int
 ExecuteCommand(cmd, really)
 char *cmd;
 int really;
@@ -2484,6 +2560,7 @@ int really;
     return(0);
 }
 
+void
 MkTmpFileName(name)
 char *name;
 {
@@ -2640,6 +2717,7 @@ char *s;
 }
 #endif
 
+void
 strcatquoting(s1, s2)
 char *s1;
 char *s2;
@@ -2655,12 +2733,12 @@ char *s2;
 #endif
 }
 
-PauseForUser() {
+void PauseForUser() {
 #if defined(MSDOS) || defined(AMIGA)
     char Buf[100];
 #endif
 
-    if (DefinitelyNotTty || MustNotBeTty) return;	
+    if (DefinitelyNotTty || MustNotBeTty) return;
 #if defined(MSDOS) || defined(AMIGA)
     printf("Press RETURN to go on\n");
     gets(Buf);
@@ -2679,23 +2757,17 @@ PauseForUser() {
 #endif
 }
 
+int
 StartRawStdin() {
 #if !defined(AMIGA) && !defined(MSDOS)
-#ifdef SYSV
-    struct termio   orterm, fterm;
-    ioctl(0, TCGETA, &orterm);	/* get current (i.e. cooked) termio */
+    struct termios   orterm, fterm;
+    tcgetattr(0, &orterm);	/* get current (i.e. cooked) termio */
     fterm = orterm;		/* get termio to modify */
 
     fterm.c_lflag &= ~ICANON;	/* clear ICANON giving raw mode */
     fterm.c_cc[VMIN] = 1;	/* set MIN char count to 1 */
     fterm.c_cc[VTIME] = 0;	/* set NO time limit */
-    return ioctl(0, TCSETAW, &fterm);	/* modify termio for raw mode */
-#else
-    struct sgttyb ts;
-    gtty(fileno(stdin), &ts);
-    ts.sg_flags |= RAW;
-    return stty(fileno(stdin), &ts);
-#endif
+    return tcsetattr(0, TCSADRAIN, &fterm);	/* modify termio for raw mode */
 #else
     return(-1);
 #endif
